keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,766 @@
+Therefore, this section describes the behavior.
+
+In summary, lien will block the namespace deletion, but would not block deletion of any resources in the namespace individually.
+
+If a lien is set to a namespace, only a deletion request to the namespace is blocked.
+Users can still request to delete each object in the namespace directly, but they can't request to delete it through the namespace deletion.
+
+If a lien is set to one of objects in a namespace, only a deletion request to the object is blocked until all other resources without liens are deleted.
+This means a resource with lien will eventually be deleted.
+Users can still request to delete its namespace or other objects in the namespace.
+
+TODO: Update below description depending on the decision.
+
+```
+When deletion of the namespace is requested, the resource deletion in the deleted namespace is done by namespace lifecycle controller in below two steps:
+1. Delete all resources (as a result, only resources without lien are deleted)
+2. Force delete all resources (as a result, all resources are deleted regardless of lien)
+
+It is to ensure proper order of deletion of resources as much as possible, even when they are force deleted through namespace deletion.
+```
+
+#### Block Adding Additional `Liens` while Deleting
+
+`Liens` shouldn't be added after `DeletionTimestamp` is non-nil, which means pre-deletion processes for finalizers are being handled.
+Adding `Liens` while deleting itself won't do any harm for the cluster, because the resource will be deleted without any additional deletion requests that will be blocked by `Liens`.
+However, users may think that the successful addition of the `Liens` means that the resource won't be deleted until the `Liens` are deleted, which isn't true.
+To avoid such a misunderstanding from users, a request for adding additional `Liens` to a resource with non-nil `DeletionTimestamp` should be blocked by api-server.
+
+#### Format of `Liens`
+
+As a result of discussions, format of `Liens` is decided as a slice of strings and it needs to be used per controller(per user) basis.
+The main reasons are scalability and better aligned with `Finalizers`.
+
+- Per controller(Per user) `Liens`
+  - Pros:
+    - Aligned with `Finalizers`
+    - Good scalability (Size of the field and number of API calls required)
+  - Cons:
+    - Bad UX (Users won't know which other objects are actually using the object)
+    - Complex consuming controllers' implementation (Each controller needs to track the using objects)
+- Per reason(Per object) `Liens`
+  - Pros:
+    - Good UX (Users can know which other objects are actually using the object)
+    - Simple consuming controllers' implementation (Lien tracks the using objects)
+  - Cons:
+    - Not aligned with `Finalizers`
+    - Bad scalability (Size of the field and number of API calls required)
```

Comment:
This could belong in the Alternatives section.

Reply:
I just deleted this section, because there is no discussion around it, recently. I can add it again if needed.